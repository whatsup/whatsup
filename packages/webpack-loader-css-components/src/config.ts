import { cosmiconfig } from 'cosmiconfig'
import { ProcessOptions, PluginCreator, Plugin } from 'postcss'
import { LoaderContext } from './index'

type PluginsConfig = { [k: string]: boolean | PluginOptions } | (string | [string, PluginOptions])[]

type PluginOptions = {
    [k: string]: any
}

type Config = {
    plugins?: PluginsConfig
    processOptions?: ProcessOptions
}

const loadRawConfig = async (loaderContext: LoaderContext) => {
    const explorer = cosmiconfig('postcss')

    try {
        const found = await explorer.search()

        if (found === null || found.isEmpty) {
            throw 'Config is empty'
        }

        const { config, filepath } = found

        loaderContext.addBuildDependency(filepath)
        loaderContext.addDependency(filepath)

        if (typeof config === 'function') {
            const ctx = {
                env: loaderContext.mode,
                mode: loaderContext.mode,
                file: loaderContext.resourcePath,
                webpackLoaderContext: loaderContext,
                options: {},
            }

            return config(ctx) as Config
        }

        return config as Config
    } catch (e) {
        return {} as Config
    }
}

function* readPlugins(plugins: PluginsConfig): Generator<[string, PluginOptions], void, unknown> {
    if (Array.isArray(plugins)) {
        for (const item of plugins) {
            if (Array.isArray(item)) {
                const [plugin, options = {}] = item

                yield [plugin, options]
            } else {
                const plugin = item
                const options = {} as PluginOptions

                yield [plugin, options]
            }
        }
    } else {
        for (const [plugin, val] of Object.entries(plugins)) {
            if (!!val) {
                const options = typeof val === 'object' ? val : {}

                yield [plugin, options]
            }
        }
    }
}

const ignorePlugins = ['postcss-import', 'postcss-url', 'postcss-modules']

const preparePlugins = (plugins: PluginsConfig) => {
    const result = [] as Plugin[]

    for (const [plugin, options] of readPlugins(plugins)) {
        if (!ignorePlugins.includes(plugin)) {
            const module = require(plugin) as PluginCreator<PluginOptions>

            result.push(module(options) as Plugin)
        }
    }

    return result
}

const prepareProcessOptions = (options: ProcessOptions, loaderContext: LoaderContext) => {
    options.from = loaderContext.resourcePath
    options.to = undefined

    return options
}

export const loadConfig = async (loaderContext: LoaderContext) => {
    const raw = await loadRawConfig(loaderContext)
    const plugins = preparePlugins(raw.plugins || [])
    const processOptions = prepareProcessOptions(raw.processOptions || {}, loaderContext)

    return { plugins, processOptions }
}
