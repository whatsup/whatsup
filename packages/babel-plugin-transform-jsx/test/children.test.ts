import { transform } from './config/transform'

describe('transform children', function () {
    it('Should transform without children', function () {
        expect(transform('<div></div>')).toMatchSnapshot()
    })

    it('Should transform once child', function () {
        expect(transform('<div><span /></div>')).toMatchSnapshot()
    })

    it('Should transform many children', function () {
        expect(transform('<ul><li /><li /></ul>')).toMatchSnapshot()
    })

    it('Should transform children with primitives', function () {
        expect(transform('<div><span>foo</span>bar<i>baz</i></div>')).toMatchSnapshot()
    })

    it('Should transform children spread', function () {
        expect(transform('<div><span>foo</span>{...other}</div>')).toMatchSnapshot()
    })

    it('Should transform children with expressions', function () {
        expect(transform('<div><span>foo</span>{[`one`, `two`].map(i => <span>{i}</span>)}</div>')).toMatchSnapshot()
    })

    it('Should trim whitespaces right only', function () {
        expect(transform('<div>{count} items </div>')).toMatchSnapshot()
    })

    it('Should trim whitespaces left only', function () {
        expect(transform('<div> items {count}</div>')).toMatchSnapshot()
    })

    it('Should trim whitespaces and create empty children', function () {
        expect(transform('<div>    </div>')).toMatchSnapshot()
    })
})
