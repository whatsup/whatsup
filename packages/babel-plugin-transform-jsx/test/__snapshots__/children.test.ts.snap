// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform children Should transform children spread 1`] = `
"import { html as _html2 } from \\"@fract/jsx\\";
import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, undefined, undefined, [_html2(\\"span\\", \\"UniqueId\\", undefined, undefined, undefined, [\\"foo\\"]), ...other]);"
`;

exports[`transform children Should transform children with expressions 1`] = `
"import { html as _html3 } from \\"@fract/jsx\\";
import { html as _html2 } from \\"@fract/jsx\\";
import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, undefined, undefined, [_html2(\\"span\\", \\"UniqueId\\", undefined, undefined, undefined, [\\"foo\\"]), [\`one\`, \`two\`].map(i => _html3(\\"span\\", \\"UniqueId\\", undefined, undefined, undefined, [i]))]);"
`;

exports[`transform children Should transform children with primitives 1`] = `
"import { html as _html3 } from \\"@fract/jsx\\";
import { html as _html2 } from \\"@fract/jsx\\";
import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, undefined, undefined, [_html2(\\"span\\", \\"UniqueId\\", undefined, undefined, undefined, [\\"foo\\"]), \\"bar\\", _html3(\\"i\\", \\"UniqueId\\", undefined, undefined, undefined, [\\"baz\\"])]);"
`;

exports[`transform children Should transform many children 1`] = `
"import { html as _html3 } from \\"@fract/jsx\\";
import { html as _html2 } from \\"@fract/jsx\\";
import { html as _html } from \\"@fract/jsx\\";

_html(\\"ul\\", \\"UniqueId\\", undefined, undefined, undefined, [_html2(\\"li\\", \\"UniqueId\\", undefined, undefined, undefined, undefined), _html3(\\"li\\", \\"UniqueId\\", undefined, undefined, undefined, undefined)]);"
`;

exports[`transform children Should transform once child 1`] = `
"import { html as _html2 } from \\"@fract/jsx\\";
import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, undefined, undefined, [_html2(\\"span\\", \\"UniqueId\\", undefined, undefined, undefined, undefined)]);"
`;

exports[`transform children Should transform without children 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, undefined, undefined, undefined);"
`;
